### PR TITLE
feat: Allow kafka consumer group id prefix configuration

### LINKED ISSUE #5063: Ability to add custom prefix to consumer ID for kafka topics in v2

### PR DESCRIPTION
**Pull request by cliveseldon**
**What this PR does / why we need it**:
Allows for customization of the Kafka consumer group id prefix.

**Which issue(s) this PR fixes**:
Fixes #5063

### CODE REVIEW

**Review by agrski — COMMENTED** [review]
Some recommended changes, although this looks to be functionally correct

File: docs/source/contents/getting-started/configuration/index.md [hunk 1]
```diff
@@ -44,6 +47,24 @@ helm install seldon-v2-runtime k8s/helm-charts/seldon-core-v2-runtime \
...
+
+If you use a shared Kafka cluster with other applications you may want to isolate the topic names and consumer group IDs from other users of the cluster to ensure there is no name clash. For this we provide two settings:
+
+ * topicPrefix : set a prefix for all topics
+ * consumerGroupIdPrefix : set a prefix for all consumer groups
+
+An example to set this in the configuration when using the helm installation is showm below:
+
+```
+helm upgrade --install seldon-v2 k8s/helm-charts/seldon-core-v2-setup/ -n seldon-mesh \
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply docs/source/contents/getting-started/configuration/index.md:60]: Will add a sentence for this in docs.

File: scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/Main.kt [hunk 2]
```diff
@@ -83,6 +83,7 @@ object Main {
             config[Cli.upstreamHost],
             config[Cli.upstreamPort],
             GrpcServiceConfigProvider.config,
+            config[Cli.kafkaConsumerGroupIdPrefix],
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/Main.kt:86]: Leaving outside as kafka properties just has actual kafka properties

File: scheduler/pkg/kafka/config/config.go [hunk 3]
```diff
@@ -25,12 +25,13 @@ import (
...
-	Producer         kafka.ConfigMap `json:"producer,omitempty"`
-	Streams          kafka.ConfigMap `json:"streams,omitempty"`
-	TopicPrefix      string          `json:"topicPrefix,omitempty"`
+	BootstrapServers      string          `json:"bootstrap.servers,omitempty"`
+	Debug                 string          `json:"debug,omitempty"`
+	Consumer              kafka.ConfigMap `json:"consumer,omitempty"`
+	Producer              kafka.ConfigMap `json:"producer,omitempty"`
+	Streams               kafka.ConfigMap `json:"streams,omitempty"`
+	TopicPrefix           string          `json:"topicPrefix,omitempty"`
+	ConsumerGroupIdPrefix string          `json:"consumerGroupIdPrefix,omitempty"`
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/kafka/config/config.go:34]: yes its a kafka.configMap and this is not a kafka config property

File: scheduler/pkg/util/hash.go [hunk 4]
```diff
@@ -50,7 +50,19 @@ func modelIdToConsumerBucket(modelId string, numBuckets int) uint32 {
 	return hash % uint32(numBuckets)
 }
 
-func GetKafkaConsumerName(modelName, prefix string, maxConsumers int) string {
+func GetKafkaConsumerName(namespace, consumerGroupIdPrefix, modelName, prefix string, maxConsumers int) string {
 	idx := modelIdToConsumerBucket(modelName, maxConsumers)
-	return fmt.Sprintf("%s-%d", prefix, idx)
+	if consumerGroupIdPrefix == "" {
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/util/hash.go:55]: ok

File: scheduler/pkg/kafka/pipeline/consumer_manager.go [hunk 5]
```diff
@@ -132,6 +135,18 @@ func (cm *ConsumerManager) Stop() {
 	}
 }
 
-func getKafkaConsumerName(prefix, id string) string {
-	return fmt.Sprintf("%s-%s", prefix, id)
+func getKafkaConsumerName(namespace, consumerGroupIdPrefix, componentPrefix, id string) string {
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/kafka/pipeline/consumer_manager.go:138]: ok will rewrite with string buffer

File: operator/apis/mlops/v1alpha1/seldonconfig_types_test.go [hunk 6]
```diff
@@ -139,7 +142,8 @@ func TestSeldonConfigurationAddDefaults(t *testing.T) {
 			},
 			expected: SeldonConfiguration{
 				KafkaConfig: KafkaConfig{
-					BootstrapServers: "h1,h2",
+					BootstrapServers:      "h1,h2",
+					ConsumerGroupIdPrefix: "foo",
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply operator/apis/mlops/v1alpha1/seldonconfig_types_test.go:146]: I modified one

File: scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/kafka/Configuration.kt [hunk 7]
```diff
@@ -156,11 +156,26 @@ fun getKafkaProperties(params: KafkaStreamsParams): KafkaProperties {
     }
 }
 
-fun KafkaProperties.withAppId(name: String): KafkaProperties {
+fun KafkaProperties.withAppId(namespace: String, consumerGroupIdPrefix: String, name: String): KafkaProperties {
     val properties = KafkaProperties()
 
     properties.putAll(this.toMap())
-    properties[StreamsConfig.APPLICATION_ID_CONFIG] = "seldon-dataflow-$name"
+    if (consumerGroupIdPrefix.equals(""))
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/kafka/Configuration.kt:163]: will change

File: scheduler/pkg/kafka/pipeline/consumer_manager_test.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+		expected              string
+	}
+	tests := []test{
+		{
+			name:                  "all params no namespace",
+			namespace:             "",
+			consumerGroupIdPrefix: "foo",
+			componentPrefix:       "pipeline",
+			id:                    "id",
+			expected:              fmt.Sprintf("%s-%s-%s", "foo", "pipeline", "id"),
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/kafka/pipeline/consumer_manager_test.go:28]: will change

File: scheduler/pkg/kafka/pipeline/consumer_manager_test.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+			consumerGroupIdPrefix: "",
+			componentPrefix:       "pipeline",
+			id:                    "id",
+			expected:              fmt.Sprintf("%s-%s-%s", "default", "pipeline", "id"),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			g.Expect(getKafkaConsumerName(test.namespace, test.consumerGroupIdPrefix, test.componentPrefix, test.id)).To(Equal(test.expected))
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/kafka/pipeline/consumer_manager_test.go:58]: will do

File: scheduler/pkg/util/hash_test.go [hunk 10]
```diff
@@ -56,3 +57,60 @@ func TestConsistentHash(t *testing.T) {
...
+		consumerGroupIdPrefix string
+		modelName             string
+		prefix                string
+		maxConsumers          int
+		expected              string
+	}
+	tests := []test{
+		{
+			name:                  "all params",
+			consumerGroupIdPrefix: "foo",
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/util/hash_test.go:76]: will change name

File: scheduler/pkg/util/hash_test.go [hunk 11]
```diff
@@ -56,3 +57,60 @@ func TestConsistentHash(t *testing.T) {
...
+			modelName:             "model",
+			prefix:                "p",
+			maxConsumers:          100,
+			expected:              fmt.Sprintf("%s-%s-%d", "foo", "p", modelIdToConsumerBucket("model", 100)),
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/util/hash_test.go:80]: will change

File: docs/source/contents/getting-started/configuration/index.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (cliveseldon) [reply docs/source/contents/getting-started/configuration/index.md:60]: Fixed in https://github.com/SeldonIO/seldon-core/pull/5072/commits/020caf4a6bd45a7b7726613162875f8765dcc406

File: operator/apis/mlops/v1alpha1/seldonconfig_types_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (cliveseldon) [reply operator/apis/mlops/v1alpha1/seldonconfig_types_test.go:147]: https://github.com/SeldonIO/seldon-core/pull/5072/commits/020caf4a6bd45a7b7726613162875f8765dcc406

File: scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/Main.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/Main.kt:86]: https://github.com/SeldonIO/seldon-core/pull/5072/commits/020caf4a6bd45a7b7726613162875f8765dcc406

File: scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/kafka/Pipeline.kt [hunk 12]
```diff
@@ -77,15 +78,17 @@ class Pipeline(
 
     companion object {
         private val logger = noCoLogger(Pipeline::class)
+        private val namespace = System.getenv("POD_NAMESPACE")
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/kafka/Pipeline.kt]: https://github.com/SeldonIO/seldon-core/pull/5072/commits/020caf4a6bd45a7b7726613162875f8765dcc406

File: scheduler/pkg/kafka/config/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/kafka/config/config.go:34]: https://github.com/SeldonIO/seldon-core/pull/5072/commits/020caf4a6bd45a7b7726613162875f8765dcc406

File: scheduler/pkg/util/hash.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/util/hash.go]: https://github.com/SeldonIO/seldon-core/pull/5072/commits/020caf4a6bd45a7b7726613162875f8765dcc406

File: scheduler/pkg/kafka/pipeline/consumer_manager.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/kafka/pipeline/consumer_manager.go:139]: https://github.com/SeldonIO/seldon-core/pull/5072/commits/020caf4a6bd45a7b7726613162875f8765dcc406

File: scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/kafka/Configuration.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/kafka/Configuration.kt]: https://github.com/SeldonIO/seldon-core/pull/5072/commits/020caf4a6bd45a7b7726613162875f8765dcc406

File: scheduler/pkg/kafka/pipeline/consumer_manager_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/kafka/pipeline/consumer_manager_test.go]: https://github.com/SeldonIO/seldon-core/pull/5072/commits/020caf4a6bd45a7b7726613162875f8765dcc406

File: scheduler/pkg/kafka/pipeline/consumer_manager_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/kafka/pipeline/consumer_manager_test.go]: https://github.com/SeldonIO/seldon-core/pull/5072/commits/020caf4a6bd45a7b7726613162875f8765dcc406

File: scheduler/pkg/util/hash_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/util/hash_test.go:75]: https://github.com/SeldonIO/seldon-core/pull/5072/commits/020caf4a6bd45a7b7726613162875f8765dcc406

File: scheduler/pkg/util/hash_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/util/hash_test.go]: https://github.com/SeldonIO/seldon-core/pull/5072/commits/020caf4a6bd45a7b7726613162875f8765dcc406

File: scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/mtls/K8sCertSecretsProvider.kt [hunk 13]
```diff
@@ -31,7 +31,7 @@ import java.nio.file.Files
 object K8sCertSecretsProvider {
 
     private val kubeConfigPath: String = System.getenv("HOME") + "/.kube/config"
-    private val namespace = System.getenv("POD_NAMESPACE")
+    private val namespace = System.getenv("SELDON_POD_NAMESPACE")
```
Comment (agrski) [inline scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/mtls/K8sCertSecretsProvider.kt:34]: :thought_balloon: I'll put in a small follow-up PR to move these to use the CLI-provided value

File: scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/PipelineSubscriber.kt [hunk 14]
```diff
@@ -72,7 +74,7 @@ class PipelineSubscriber(
     suspend fun subscribe() {
         logger.info("will connect to ${upstreamHost}:${upstreamPort}")
         retry(grpcFailurePolicy + binaryExponentialBackoff(50..5_000L)) {
-            subscribePipelines()
+            subscribePipelines(kafkaConsumerGroupIdPrefix, namespace)
```
Comment (agrski) [inline scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/PipelineSubscriber.kt:77]: :upside_down_face: No need to pass these parameters explicitly as they've `private val`s above; happy to change in a follow-up PR for expediency

**Review by agrski — APPROVED** [review]
Only minor comments outstanding; approving and will create a follow-up for a few tidy-ups as indicated in comments

File: scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/kafka/Pipeline.kt [hunk 15]
```diff
@@ -25,6 +25,7 @@ import org.apache.kafka.streams.StreamsBuilder
 import org.apache.kafka.streams.StreamsConfig
 import org.apache.kafka.streams.Topology
 import java.util.concurrent.CountDownLatch
+import javax.xml.stream.events.Namespace
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/kafka/Pipeline.kt:28]: will remove

**Pull request merged**